### PR TITLE
feat(api): show verifications in API

### PR DESCRIPTION
**Pull request by lorin**
# Problem

Provide the UI with information about the state of verifications for artifact versions in each environment. (For example, what verifications have been run against `keel-0.850.0-h1020.a525f99` in the `prestaging` environment)?

# Implemented solution

In the `/application/{application}` endpoint (called by deck in the environments view), add additional fields to artifact version objects which encode verification summary notification.

In *jq* notation, these verification summaries are found in: `artifacts[].versions[].environments[].verifications`.

## Example output

For the `fnord` app, the API call would be: 

http://gate/managed/application/fnord?entities=resources&entities=artifacts&entities=environments

The  output would looke like this:

```json
{
  "artifacts": [
    {
      "name": "fnord",
      "type": "deb",
      "reference": "fnord",
      "versions": [
        {
          "version": "fnord-0.850.0-h1021.b527f39",
          "environments": [
            {
              "name": "staging",
              "verifications": [
                {
                  "id": "test-container:acme/testcon:stable",
                  "type": "test-container",
                  "status": "PASSED",
                  "startedAt": "2021-01-04T22:31:12.142Z",
                  "completedAt": "2021-01-04T22:52:07.358Z",
                  "link": "http://www.example.com/test-output/abc123"
                }
              ]
            }
          ]
```

## Guide to main code changes

ApplicationService.kt - this is the meat of the change. (to be completed)

ArtifactSummary.kt - the new `ArtifactVersionEnvironmentVerificationSummary` data class maps to the payload in the `verifications` field shown above. 

Verification.kt - Verification interface now has a `link` function for generating a URL where users can learn more about the current state of a verification. This hasn't been implemented yet for test-container verifications (see future work).

The other changes are just plumbing and tests.

## Future work

[ ] Generate the `link` URL to point to the location where container output can be found.

### CODE REVIEW

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/persistence/KeelReadOnlyRepository.kt [hunk 1]
```diff
@@ -58,6 +60,8 @@ interface KeelReadOnlyRepository {
 
   fun artifactVersions(artifact: DeliveryArtifact, limit: Int = DEFAULT_MAX_ARTIFACT_VERSIONS): List<PublishedArtifact>
 
+  fun getStatesBatch(contexts: List<VerificationContext>) : List<Map<String, VerificationState>>
```
Comment (luispollo) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/persistence/KeelReadOnlyRepository.kt:63]:
```suggestion
  fun getVerificationStatesBatch(contexts: List<VerificationContext>) : List<Map<String, VerificationState>>
```

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/verification/VerificationRepository.kt [hunk 2]
```diff
@@ -26,6 +26,23 @@ interface VerificationRepository {
...
+   *    val contexts : List<VerificationContext> = ...
+   *    contexts.map { context -> this.getStates(context) }
+   *
+   * It exists as a separate call because it can be much more efficient to query the underlying repository as a batch.
+   *
+   * @param contexts a list of verification contexts to query fo rstate
+   *
+   * @return a list of maps of verification ids to states, in the same order as the contexts
+   */
+  fun getStatesBatch(contexts: List<VerificationContext>) : List<Map<String, VerificationState>>
```
Comment (luispollo) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/api/verification/VerificationRepository.kt:44]:
```suggestion
  fun getVerificationStatesBatch(contexts: List<VerificationContext>) : List<Map<String, VerificationState>>
```

File: keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/VerificationRepositoryTests.kt [hunk 3]
```diff
@@ -377,4 +382,101 @@ abstract class VerificationRepositoryTests<IMPLEMENTATION : VerificationReposito
...
+
+  //
+  // getStatesBatch tests
+  //
+  @Test
+  fun `no verification states`() {
+    context.setup()
+    val contexts = listOf(context)
+
+    expectThat(subject.getStatesBatch(contexts))
```
Comment (luispollo) [inline keel-core-test/src/main/kotlin/com/netflix/spinnaker/keel/persistence/VerificationRepositoryTests.kt:394]:
```suggestion
    expectThat(subject.getVerificationStatesBatch(contexts))
```

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepository.kt [hunk 4]
```diff
@@ -430,5 +433,9 @@ class CombinedRepository(
     limit: Int
   ) : Collection<VerificationContext> =
     verificationRepository.nextEnvironmentsForVerification(minTimeSinceLastCheck, limit)
+
+  override fun getStatesBatch(contexts: List<VerificationContext>) : List<Map<String, VerificationState>> =
```
Comment (luispollo) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/CombinedRepository.kt:437]:
```suggestion
  override fun getVerificationStatesBatch(contexts: List<VerificationContext>) : List<Map<String, VerificationState>> =
```

**Comment by lorin** [comment]
Closing this draft PR, real PRs coming today.

**Pull request closed without merging**